### PR TITLE
Add auth header for build

### PR DESCRIPTION
**Pull request by BastienAr**
Docker build offers a way to pass auth configurations through `X-Registry-Config` header. I've just implemented a way to use that feature with docker-client.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,7 @@
 npipe.exe
 .DS_Store
 out/
+*.settings
+*.project
+*bin/
+*.classpath
```
Comment (gesellix) [inline .gitignore]: Please add a newline here.

File: client/src/main/groovy/de/gesellix/docker/client/image/ManageImageClient.groovy [hunk 2]
```diff
@@ -76,16 +76,21 @@ class ManageImageClient implements ManageImage {
...
-                                    body              : buildContext,
-                                    requestContentType: "application/octet-stream",
-                                    async             : async])
+        buildOptions = buildOptions ?: [:]
+        def request = [path              : "/build",
+                       query             : actualQuery,
+                       body              : buildContext,
+                       requestContentType: "application/octet-stream",
+                       async             : async]
+        if (buildOptions.EncodedRegistryConfig) {
```
Comment (gesellix) [inline client/src/main/groovy/de/gesellix/docker/client/image/ManageImageClient.groovy]: I'm aware of the different header names for the same use case, but I would prefer to have a consistent Java api (I'm slowly working towards a more type safe approach). So, please rename the build option `EncodedRegistryConfig` to `EncodedRegistryAuth`, please.

Comment (BastienAr) [reply client/src/main/groovy/de/gesellix/docker/client/image/ManageImageClient.groovy]: Ok, no problem. I will check for the integration test tonight or tomorrow. Thanks for your review

**Comment by BastienAr** [comment]
I've fixed the integration test, but I'm afraid this will introduce an API break (because you need to send an empty - or not - map in order to use the `DockerAsyncCallback`). Maybe this is fine. If not, I can set the options arg at the end of the ` de.gesellix.docker.client.image.ManageImage.build` method args.

Please also note that I reversed the `EncodedRegistryConfig` to `EncodedRegistryAuth` but it's not the same object. Docker API doc says: 

- **EncodedRegistryAuth**
```json
{
  "username": "string",
  "password": "string",
  "email": "string",
  "serveraddress": "string"
}
```

- **EncodedRegistryConfig**
```json
{
  "docker.example.com": {
    "username": "janedoe",
    "password": "hunter2"
  },
  "https://index.docker.io/v1/": {
    "username": "mobydock",
    "password": "conta1n3rize14"
  }
}
```

It seems like you can set multiple repos auth for docker build (probably to be multi-stage compliant).

**Comment by gesellix** [comment]
@BastienAr oh, I'm sorry, I didn't recognize the different content for `EncodedRegistry<Auth|Config>`. Then we'll need different names. If that's not too much hassle for you, I would ask you to revert the property renaming.

Regarding the breaking api change: I suggest to introduce a parameter object `BuildConfig` containing all currently available method parameters for `build...()`. A new `build(BuildConfig config)` method could then be added beside the currently available methods. Would that work for you?

**Comment by BastienAr** [comment]
Yep, I think I can find the time to rewrite that today

**Comment by gesellix** [comment]
Many thanks for your contribution and your patience with my misunderstanding of `config` vs `auth` :)
I'll merge this now, but I'll also take some time to think about the changed api. I'll probably declare some old methods as deprecated before cutting a release.

**Comment by BastienAr** [comment]
It would be could to integrate it to the gradle plugin, I can check that when the docker client is released

**Comment by gesellix** [comment]
@BastienAr please have a look at https://github.com/gesellix/docker-client/commit/b182e385175cf7c086cff4eae6c71f88d6dd2eeb - I further refactored the `build*` methods at the cost of a breaking change for the specific `build(InputStream)` method. That allowed me to also make the return type to be specific (with the newly introduced `BuildResult`). Thanks again for your effort and more or less pushing me to clean up that old stuff :)

The new release is available as `'de.gesellix:docker-client:2018-01-26T21-07-56'`.

I'll take care of the Gradle plugin tonight (the breaking change requires a coordinated announcement, I guess.).

**Comment by gesellix** [comment]
P.S. the Gradle plugin has been updated, too. If you like, you can add the build auth config now.

**Pull request merged**